src/main/java/org/gradle/profiler/BuildActionInvoker.java

```diff
@@ -0,0 +1,68 @@
+package org.gradle.profiler;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+import static org.gradle.profiler.Logging.startOperation;
+
+public class BuildActionInvoker {
+    private final List<String> jvmArgs;
+    private final List<String> gradleArgs;
+    private final PidInstrumentation pidInstrumentation;
+    private final Consumer<BuildInvocationResult> resultsConsumer;
+    private final BuildInvoker buildInvoker;
+
+    public BuildActionInvoker(List<String> jvmArgs, List<String> gradleArgs, BuildInvoker buildInvoker, PidInstrumentation pidInstrumentation, Consumer<BuildInvocationResult> resultsConsumer) {
+        this.jvmArgs = jvmArgs;
+        this.gradleArgs = gradleArgs;
+        this.buildInvoker = buildInvoker;
+        this.pidInstrumentation = pidInstrumentation;
+        this.resultsConsumer = resultsConsumer;
+    }
+
+    public BuildInvocationResult runBuild(Phase phase, int buildNumber, BuildStep buildStep, List<String> tasks, BuildAction buildAction) {
```

Comment:
I think it would be clearer if the build action contained the tasks it will run instead of taking them as an argument.

Reply:
Yes, I thought so too, but was going to do this as a follow-up PR. I can change it in this PR before merging if you'd prefer.

Reply:
Nothing that holds up this PR.